### PR TITLE
Change DataCatalog interface

### LINKED ISSUE #1007: Change DataCatalog interface
Change DataCatalog interface and the existing GrpcDataCatalog implementation. This deals with generalisation of DataCatalog interface to adapt to openapi based format.

### PR DESCRIPTION
**Pull request by rohithdv**
This PR fixes #1007. 

This PR has the following changes:

- [x] Change to the DataCatalog interface to use openapi objects instead of grpc objects.
- [x] Changes to GRPC DataCatalog implementation to account for the interface changes
- [x] Addition of conversion functions to/from GRPC objects from.to OpenApi objects.
- [x] Reorganize `pkg/connectors/clients` folder into `pkg/connectors/datacatalog/clients` and `pkg/connectors/policymanager/clients` for consistency.
- [x] Changes in some places in manager test code to account for the interface change.
- [x] Add openapi implementation of Datacatalog interface
- [x] Fixes to update opa connector to use the openapi catalog response
- [x] Add check in manager to use openapi datacatalog if connector url startswith http, otherwise use the grpc based datacatalog.

Signed-off-by: rohithdv rovallam@in.ibm.com

### CODE REVIEW

File: connectors/opa/lib/catalog_reader.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 
 	clients "fybrik.io/fybrik/pkg/connectors/clients"
 	pb "fybrik.io/fybrik/pkg/connectors/protobuf"
+	datacatalogTaxonomyModels "fybrik.io/fybrik/pkg/taxonomy/model/datacatalog/base"
 	taxonomymodels "fybrik.io/fybrik/pkg/taxonomy/model/policymanager/base"
```
Comment (shlomitk1) [inline connectors/opa/lib/catalog_reader.go:15]: rename to policyTaxonomyModels for consistency

File: connectors/opa/lib/catalog_reader.go [hunk 2]
```diff
@@ -28,8 +29,11 @@ func (r *CatalogReader) GetDatasetsMetadataFromCatalog(in *taxonomymodels.Policy
 	datasetsMetadata := make(map[string]interface{})
 	datasetID := (in.GetResource()).Name
 	if _, present := datasetsMetadata[datasetID]; !present {
-		objToSend := &pb.CatalogDatasetRequest{CredentialPath: creds, DatasetId: datasetID}
-		info, err := (*r.DataCatalog).GetDatasetInfo(context.Background(), objToSend)
+		// objToSend := &pb.CatalogDatasetRequest{CredentialPath: creds, DatasetId: datasetID}
```
Comment (shlomitk1) [inline connectors/opa/lib/catalog_reader.go:32]: Remove commented lines that refer to old code

File: connectors/opa/lib/opa_reader.go [hunk 3]
```diff
@@ -8,8 +8,8 @@ import (
 	"fmt"
 	"log"
 	"net/http"
-	"strings"
 
+	datacatalogTaxonomyModels "fybrik.io/fybrik/pkg/taxonomy/model/datacatalog/base"
 	taxonomymodels "fybrik.io/fybrik/pkg/taxonomy/model/policymanager/base"
```
Comment (shlomitk1) [inline connectors/opa/lib/opa_reader.go:13]: policyTaxonomyModels?

File: connectors/opa/lib/opa_reader.go [hunk 4]
```diff
@@ -22,130 +22,195 @@ func NewOpaReader(opasrvurl string, client *http.Client) *OpaReader {
 	return &OpaReader{opaServerURL: opasrvurl, opaClient: client}
 }
 
-func (r *OpaReader) updatePolicyManagerRequestWithResourceInfo(in *taxonomymodels.PolicyManagerRequest, catalogMetadata map[string]interface{}) (*taxonomymodels.PolicyManagerRequest, error) {
-	responseBytes, errJSON := json.MarshalIndent(catalogMetadata, "", "\t")
+func (r *OpaReader) updatePolicyManagerRequestWithResourceInfo2(in *taxonomymodels.PolicyManagerRequest, catalogMetadata *datacatalogTaxonomyModels.DataCatalogResponse) (*taxonomymodels.PolicyManagerRequest, error) {
```
Comment (shlomitk1) [inline connectors/opa/lib/opa_reader.go:25]: Why 2 in updatePolicyManagerRequestWithResourceInfo2?

File: connectors/opa/lib/opa_reader.go [hunk 5]
```diff
@@ -22,130 +22,195 @@ func NewOpaReader(opasrvurl string, client *http.Client) *OpaReader {
...
+	if errJSON != nil {
+		return nil, fmt.Errorf("error Marshalling taxonomymodels.PolicyManagerRequest in updatePolicyManagerRequestWithResourceInfo2: %v", errJSON)
 	}
+	log.Print("returning updated taxonomymodels.PolicyManagerRequest in updatePolicyManagerRequestWithResourceInfo2:" + string(responseBytes1))
+	// just printing - end
+
 	return in, nil
 }
 
+// func (r *OpaReader) updatePolicyManagerRequestWithResourceInfo(in *taxonomymodels.PolicyManagerRequest, catalogMetadata map[string]interface{}) (*taxonomymodels.PolicyManagerRequest, error) {
```
Comment (shlomitk1) [inline connectors/opa/lib/opa_reader.go:50]: Remove

File: connectors/opa/lib/opa_reader.go [hunk 6]
```diff
@@ -22,130 +22,195 @@ func NewOpaReader(opasrvurl string, client *http.Client) *OpaReader {
 	return &OpaReader{opaServerURL: opasrvurl, opaClient: client}
 }
 
-func (r *OpaReader) updatePolicyManagerRequestWithResourceInfo(in *taxonomymodels.PolicyManagerRequest, catalogMetadata map[string]interface{}) (*taxonomymodels.PolicyManagerRequest, error) {
-	responseBytes, errJSON := json.MarshalIndent(catalogMetadata, "", "\t")
+func (r *OpaReader) updatePolicyManagerRequestWithResourceInfo2(in *taxonomymodels.PolicyManagerRequest, catalogMetadata *datacatalogTaxonomyModels.DataCatalogResponse) (*taxonomymodels.PolicyManagerRequest, error) {
+	// just printing - start
+	responseBytes1, errJSON := json.MarshalIndent(catalogMetadata.ResourceMetadata, "", "\t")
```
Comment (shlomitk1) [inline connectors/opa/lib/opa_reader.go:27]: Why do you have 1 and 2 in names of variables and functions?

File: connectors/opa/lib/catalog_reader.go (hunk 1 above)
Comment (roee88) [reply connectors/opa/lib/catalog_reader.go:15]: This entire file seems to be unused

File: manager/controllers/mockup/datacatalog.go [hunk 7]
```diff
@@ -46,134 +49,123 @@ func (d *DataCatalogDummy) Close() error {
...
-					ObjectKey: "test.csv",
-				},
-			},
-			CredentialsInfo: &pb.CredentialsInfo{
-				VaultSecretPath: "/v1/kubernetes-secrets/creds-secret-name?namespace=fybrik-system",
-			},
-			Metadata: &pb.DatasetMetadata{DatasetTags: []string{"PI"}},
+
+	tags := make(map[string]interface{})
+	tags["tags"] = []string{"PI"}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply manager/controllers/mockup/datacatalog.go]: fixed. Please check. Thanks.

File: connectors/opa/lib/opa_reader.go (hunk 6 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go]: fixed. Thanks.

File: connectors/opa/lib/opa_reader.go (hunk 5 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go]: fixed. Thanks.

File: connectors/opa/lib/opa_reader.go (hunk 4 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go]: fixed. Thanks.

File: connectors/opa/lib/opa_reader.go (hunk 3 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go]: fixed. Thanks.

File: connectors/opa/lib/catalog_reader.go (hunk 2 above)
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: fixed. Thanks.

File: connectors/opa/lib/catalog_reader.go (hunk 1 above)
Comment (rohithdv) [reply connectors/opa/lib/catalog_reader.go]: Removed catalog_reader and updated opa reader as many functions are not required now. Please review.

File: connectors/opa/lib/opa_reader.go [hunk 8]
```diff
@@ -8,170 +8,92 @@ import (
...
 }
 
-func (r *OpaReader) GetOPADecisions(in *taxonomymodels.PolicyManagerRequest, creds string, catalogReader *CatalogReader, policyToBeEvaluated string) (taxonomymodels.PolicyManagerResponse, error) {
-	datasetsMetadata, err := catalogReader.GetDatasetsMetadataFromCatalog(in, creds)
+func (r *OpaReader) GetOPADecisions(in *policymanagerTaxonomyModels.PolicyManagerRequest, creds string, policyToBeEvaluated string) (policymanagerTaxonomyModels.PolicyManagerResponse, error) {
+	datasetID := (in.GetResource()).Name
+	objToSend := datacatalogTaxonomyModels.DataCatalogRequest{AssetID: datasetID, OperationType: datacatalogTaxonomyModels.READ}
+
+	info, err := (*r.dataCatalog).GetAssetInfo(&objToSend, creds)
+	// info, err := (*r.DataCatalog).GetDatasetInfo(context.Background(), objToSend)
```
Comment (shlomitk1) [inline connectors/opa/lib/opa_reader.go:57]: remove this line

File: manager/controllers/app/data_catalog_interface.go [hunk 9]
```diff
@@ -25,31 +23,33 @@ import (
 // - an error if happened
 // - the new asset identifier
 func (r *FybrikApplicationReconciler) RegisterAsset(catalogID string, info *app.DatasetDetails, input *app.FybrikApplication) (string, error) {
-	datasetDetails := &pb.DatasetDetails{}
-	err := info.Details.Into(datasetDetails)
-	if err != nil {
-		return "", err
-	}
+	/*	datasetDetails := &pb.DatasetDetails{}
```
Comment (shlomitk1) [inline manager/controllers/app/data_catalog_interface.go:26]: remove the commented code

File: test/services/datacatalog/datacatalogstub.go [hunk 10]
```diff
@@ -4,33 +4,52 @@
...
+		dataCatalogResp, err := dataCatalog.GetAssetInfo(&dataCatalogReq, creds)
+		if err != nil {
+			c.String(http.StatusInternalServerError, "Error in getAssetInfo!")
+			return
+		}
+		c.JSON(http.StatusOK, dataCatalogResp)
+	})
+
+	router.GET("/", func(c *gin.Context) {
+		c.String(http.StatusOK, "Hello World!")
```
Comment (shlomitk1) [inline test/services/datacatalog/datacatalogstub.go:51]: Replace "Hello World' with something meaningful here

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+func ConvertDataCatalogGrpcRespToOpenAPIResp(result *pb.CatalogDatasetInfo) (*datacatalogTaxonomyModels.DataCatalogResponse, error) {
+	// convert GRPC response to Open Api Response - start
+	resourceCols := make([]datacatalogTaxonomyModels.ResourceColumns, 0)
+
+	for colName, compMetaData := range result.GetDetails().Metadata.GetComponentsMetadata() {
+		if compMetaData != nil {
+			rscCol := datacatalogTaxonomyModels.ResourceColumns{
+				Name: colName}
+			rsCalMap := make(map[string]interface{})
```
Comment (shlomitk1) [inline pkg/connectors/datacatalog/clients/datacatalog_grpc.go:107]: rsColMap?

File: manager/controllers/app/data_catalog_interface.go (hunk 9 above)
Comment (rohithdv) [reply manager/controllers/app/data_catalog_interface.go:26]: Are we sure about removing this ? There was an ask to support `registerAsset` after `getAssetInfo` is working. Do we want a reference in commented code here for that ? Please suggest. Thanks.

File: test/services/datacatalog/datacatalogstub.go (hunk 10 above)
Comment (rohithdv) [reply test/services/datacatalog/datacatalogstub.go:51]: used the message "Serving REST APIs as part of data catalog stub". Hope it is ok. Please review.

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go (hunk 11 above)
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_grpc.go:107]: fixed. thanks.

File: connectors/opa/lib/opa_reader.go (hunk 8 above)
Comment (rohithdv) [reply connectors/opa/lib/opa_reader.go:57]: fixed. thanks.

**Review by roee88 — APPROVED** [review]

File: manager/controllers/app/data_catalog_interface.go (hunk 9 above)
Comment (roee88) [reply manager/controllers/app/data_catalog_interface.go:26]: When the time comes if you need this older code just pick at the git history ;)

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	res, err := json.MarshalIndent(dataCatalogResp, "", "\t")
+	if err != nil {
+		log.Println("Error during marshalling data catalog response: ", err)
+		return nil, err
+	}
+	log.Println("Marshalled value of data catalog response: ", string(res))
+	return dataCatalogResp, nil
+	// return result, err
```
Comment (shlomitk1) [inline pkg/connectors/datacatalog/clients/datacatalog_grpc.go:79]: remove this line

**Review by shlomitk1 — APPROVED** [review]

File: manager/controllers/app/data_catalog_interface.go (hunk 9 above)
Comment (rohithdv) [reply manager/controllers/app/data_catalog_interface.go]: sure. fixed it now.

File: pkg/connectors/datacatalog/clients/datacatalog_grpc.go (hunk 12 above)
Comment (rohithdv) [reply pkg/connectors/datacatalog/clients/datacatalog_grpc.go]: fixed. thanks.

**Pull request merged**